include/mpi_support.h

```diff
@@ -1,56 +1,97 @@
+	constexpr int TAG_DATA_TRANSFER = 1;
+	constexpr int TAG_TELEMETRY = 2;
+
+} // namespace mpi_support
+
+struct from_payload_size_tag {
+} inline constexpr from_payload_size;
+
+struct from_frame_bytes_tag {
+} inline constexpr from_frame_bytes;
+
+// unique_frame_ptr manually `operator new`s the underlying frame memory, placement-new-constructs the frame and casts it to a frame pointer.
+// I'm convinced that I'm actually, technically allowed to use the resulting frame pointer in a delete-expression and therefore keep `std::default_delete` as
+// the deleter type for `unique_frame_ptr::impl`: Following the standard, delete-expression requires its operand to originate from a new-expression,
+// and placement-new is defined to be a new-expression. The following implicit call to operator delete is also legal, since memory was obtained from
+// `operator new`. Despite the beauty of this standards loophole, @BlackMark29A and @PeterTh couldn't be convinced to let me merge it :(   -- @fknorr
+template <typename Frame>
+struct unique_frame_delete {
+	void operator()(Frame* frame) const {
+		if(frame) {
+			frame->~Frame();
+			operator delete(frame);
+		}
+	}
+};
+
+/**
+ * Owning smart pointer for variable-sized structures with a 0-sized array of type Frame::payload_type as the last member.
+ */
+template <typename Frame>
+class unique_frame_ptr : private std::unique_ptr<Frame, unique_frame_delete<Frame>> {
+  private:
+	using impl = std::unique_ptr<Frame, unique_frame_delete<Frame>>;
+
+  public:
+	using payload_type = typename Frame::payload_type;
+
+	unique_frame_ptr() = default;
+
+	unique_frame_ptr(from_payload_size_tag, size_t payload_size) : impl(make_impl(frame_bytes_from_payload_size(payload_size))), payload_size(payload_size) {}
+
+	unique_frame_ptr(from_frame_bytes_tag, size_t frame_bytes) : impl(make_impl(frame_bytes)), payload_size(payload_size_from_frame_bytes(frame_bytes)) {}
+
+	unique_frame_ptr(unique_frame_ptr&& other) noexcept : impl(static_cast<impl&&>(other)), payload_size(other.payload_size) { other.payload_size = 0; }
+
+	unique_frame_ptr& operator=(unique_frame_ptr&& other) noexcept {
+		static_cast<impl&>(*this) = static_cast<impl&&>(other);
+		payload_size = other.payload_size;
+		other.payload_size = 0;
+		return *this;
+	}
+
+	Frame* get_pointer() { return impl::get(); }
+	const Frame* get_pointer() const { return impl::get(); }
+	size_t get_payload_size() const { return payload_size; }
+	size_t get_frame_size_bytes() const { return frame_bytes_from_payload_size(payload_size); }
+
+	using impl::operator bool;
+	using impl::operator*;
+	using impl::operator->;
+
+  private:
+	size_t payload_size = 0;
+
+	static impl make_impl(const size_t frame_bytes) {
+		const auto mem = operator new(frame_bytes);
+		try {
+			new(mem) Frame;
+		} catch(...) {
+			operator delete(mem);
+			throw;
+		}
+		return impl{static_cast<Frame*>(mem)};
+	}
+
+	static size_t frame_bytes_from_payload_size(size_t payload_size) { //
```

Comment:
Might be better to name this something like `payload_count`. `payload_size` sounds like it's in bytes.

Reply:
Done.